ui/pages/confirmations/hooks/alerts/transactions/useMultipleApprovalsAlerts.ts

```diff
@@ -0,0 +1,164 @@
+import { Alert } from '../../../../../ducks/confirm-alerts/confirm-alerts';
+import { Severity } from '../../../../../helpers/constants/design-system';
+import { useI18nContext } from '../../../../../hooks/useI18nContext';
+import { useConfirmContext } from '../../../context/confirm';
+
+type ApprovalInfo = {
+  tokenAddress: Hex;
+  spenderAddress?: Hex;
+  amount?: BigNumber;
+  isNFT: boolean;
+};
+
+type TokenOutflow = {
+  tokenAddress: Hex;
+  amount: BigNumber;
+  isDecrease: boolean;
+};
+
+export function useMultipleApprovalsAlerts(): Alert[] {
+  const t = useI18nContext();
+  const { currentConfirmation } = useConfirmContext<TransactionMeta>();
+
+  const isBatchTransaction = currentConfirmation?.type === 'batch';
+  const nestedTransactions = currentConfirmation?.nestedTransactions;
+  const simulationDataArray =
+    currentConfirmation?.simulationData?.tokenBalanceChanges;
+
+  // Extract all approvals from nested transactions
+  const approvals = useMemo((): ApprovalInfo[] => {
+    if (!isBatchTransaction || !nestedTransactions) {
+      return [];
+    }
+
+    const approvalsList: ApprovalInfo[] = [];
+
+    nestedTransactions.forEach((transaction) => {
+      const { data, to } = transaction;
+      if (!data || !to) {
+        return;
+      }
+
+      const parseResult = parseApprovalTransactionData(data);
+      if (!parseResult) {
+        return;
+      }
+
+      const { name, amountOrTokenId, tokenAddress } = parseResult;
+
+      // Skip revocations (zero amounts or setApprovalForAll with false)
+      if (amountOrTokenId?.isZero() || parseResult.isRevokeAll) {
+        return;
+      }
+
+      let actualTokenAddress: Hex;
+      let spenderAddress: Hex | undefined;
+
+      switch (name) {
+        case 'approve':
+          if (tokenAddress) {
+            // Permit2 approve
+            actualTokenAddress = tokenAddress;
+            spenderAddress = to;
+          } else {
+            // Regular ERC20 approve
+            actualTokenAddress = to;
+          }
+          break;
+        case 'increaseAllowance':
+        case 'setApprovalForAll':
+          actualTokenAddress = to;
+          break;
+        default:
+          return;
+      }
+
+      const isNFT = name === 'setApprovalForAll';
+
+      approvalsList.push({
+        tokenAddress: actualTokenAddress,
+        spenderAddress,
+        amount: amountOrTokenId,
+        isNFT,
+      });
+    });
+
+    return approvalsList;
+  }, [isBatchTransaction, nestedTransactions]);
+
+  const tokenOutflows = useMemo((): TokenOutflow[] => {
+    if (!simulationDataArray) {
+      return [];
+    }
+
+    return simulationDataArray
+      .filter((change) => change.isDecrease)
+      .map((change) => ({
+        tokenAddress: change.address.toLowerCase() as Hex,
+        amount: new BigNumber(change.difference, 16),
+        isDecrease: change.isDecrease,
+      }));
+  }, [simulationDataArray]);
+
+  const unusedApprovals = useMemo(() => {
```

Comment:
Could we break this up into multiple local utility functions to simplify to something like:

```
const nonZeroApprovalTokens = getNonZeroApprovalTokens(approvalBalanceChanges);
const tokensWithDecrease = getTokensWithDecrease(simulationData);
const hasUnnecessaryApprovals = checkApprovalsUnnecessary(nonZeroApprovalTokens, tokensWithDecrease)
```